### ISSUE TITLE
[ESLint v8] rule errors from generic eslint comment handling

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#14960]
**Tell us about your environment**

* **ESLint Version:** v8.0.0-beta.0
* **Node Version:** v16.6.0
* **npm Version:** v7.19.1
* **Operating System:** MacOS

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
// eslint comment that's not actually an eslint comment
```

```bash
yarn eslint ./test.js
```

**What did you expect to happen?**

No errors

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

```
/Users/bradzacher/testbed/test.js
  1:1  error  Failed to parse JSON from ' comment that's not actually an eslint comment': Unexpected token c in JSON at position 2
```

**Steps to reproduce this issue:**

1. create new folder
2. install eslint v8
3. add a file test.js with the above code
4. add an empty eslintrc as above
5. run eslint on the file

**Comment by mdjermanovic** [comment]
This is expected per [RFC - Allow All Directives in Line Comments ](https://github.com/eslint/rfcs/tree/main/designs/2019-allow-all-directives-in-line-comments). The solution is to change the comments (for example, we changed `// global...` to `// Global...` in PR https://github.com/eslint/eslint/pull/14656 which implemented the RFC).

**Comment by bradzacher** [comment]
So any comment starting with the words `exported`, `global`, `globals`, or `eslint` will now be parsed by ESLint and will now cause linting errors.

IMO this change needs more thought and work before the v8 release as that's a pretty unacceptable change for users for a comparably **very minor** devx win.

I see three major problems with this feature as is:

# Migration Effort

This change forces codebases to do a migration on comments. Not even on ESLint-specific comments - just on random comments throughout their codebase.

I did a quick check of the FB codebase and found _**200**_ comments which need to be updated.
Many of them are "multiline singleline" comments:
```js
// this is the start of my comment which uses the word
// global on the second line
```

So the fix to change `s/global/Global/` isn't great as it would inject a capital into the middle of the comment.
This breaking change will require time and effort to find, investigate and codemod.
This is likely going to encourage people to not bother with the upgrade in the interim.

# Assumes users read the docs

Most users will not read the ESLint docs. Especially at any medium-to-large company where linting is managed by someone else - the majority of users will be just consumers of an existing linting environment.
They won't know what the heck `Failed to parse JSON from ' comment that's not actually an eslint comment': Unexpected token c in JSON at position 2` means, nor will they want to go and read ESLint's v8 release notes to figure it out.

Similarly a user that sees all of those `no-unused-vars` errors isn't going to immediately click as to their meaning. Given today's codebases - those users have likely never seen a `/* global */` comment as globals (if any) were configured via the `.eslintrc`.

I myself only understood why it errored not because I pattern-matched it as a globals comment I've seen before - but because I remembered the globals augmentation from when I reimplemented the `no-unused-vars` rule.

# Poor messaging

This is very closely related to the last point.

"Failed to parse JSON" is a super cryptic error message to see. It means absolutely nothing to a user: "I just wrote a random comment, why is there now a JSON parse error on that comment???"

I encountered it here in typescript-eslint: https://github.com/typescript-eslint/typescript-eslint/blob/a752638f975935882746ee5ef614ca02f120675d/.eslintrc.js#L80-L82
`// eslint base` produced the error `Failed to parse JSON from ' base': Unexpected token 'b' in JSON at position 2`

Either this should produce no errors at all - or it should catch the error and explain the actual problem.

Similarly the "global" handling is **incredibly cryptic**.
```js
// this is the start of my comment which uses the word
// global on the second line
```

This innocent comment should not result in 4 `no-unused-vars` errors. Or if it does - it should clearly explain what the problem is and how to fix it.
I could also easily see cryptic behaviour if the comment contains actual globals defined elsewhere in it as well, eg:
```ts
// this is the start of my comment which uses the word
// global it describes a test written using jest

jest.spyOn(...);
it('', () => {
  describe('', () => {
    test('', () => {});
  });
});

////
//  3:14  error  'has' is defined but never used      no-unused-vars
//  3:18  error  'a' is defined but never used        no-unused-vars
//  3:25  error  'written' is defined but never used  no-unused-vars
//  3:33  error  'using' is defined but never used    no-unused-vars
// BUT it, test and jest do not get reported - what?
```

---

Sorry if the wording is a bit harsh above. I tried to boil it down to the facts. I am just easily frustrated by easily avoidable poor UX, and this is exactly that - a tiny devx win for a huge UX loss.

**Comment by aladdin-add** [comment]
the change was to improve the DX, but we have to reconsider it if it brings quite a few migrations cost.


just a thought: as the eslint evolves, at some point we may change the eslint directives to be more consistent(`eslint-xxx`), e.g.

```js
// global(s) => // eslint-global(s)
// export => // eslint-export
// eslint foo => // eslint-rule foo ?
// eslint-enable(-line) => no changes ?
// eslint-env => no changes ?
```

**Comment by mdjermanovic** [comment]
> I did a quick check of the FB codebase and found _**200**_ comments which need to be updated.
> Many of them are "multiline singleline" comments:
> 
> ```js
> // this is the start of my comment which uses the word
> // global on the second line
> ```

@eslint/eslint-tsc thoughts about this? I'd be personally fine with reverting the change, as we might have underestimated the impact.

**Comment by bradzacher** [comment]
To be clear - I think this is a great change - but I think that something like what @aladdin-add suggested is a hard pre-requisite for the feature to disambiguate comments.

----

another example that's relatively common is for people to refer to "eslint" when explaining their usage of a disable comment:

```js
// eslint doesn't understand how to foo the bar
// eslint-disable-next-line plugin/foo-bar
```

**Comment by nzakas** [comment]
I’m fine with reverting. It seems like the cost outweighs the added value. If we want to reintroduce at some point then I agree with @aladdin-add’s point that coming up with an unambiguous format is the right way to go.

**Comment by thernstig** [comment]
Could https://eslint.org/blog/2021/06/whats-coming-in-eslint-8.0.0 be updated to say this was reverted? I tried to convert our code to this just to notice it was not supported in v8.8.0.
